File: 23_设计类/设计版本控制(可持久化)/可持久化数组/PersistentQueue/Stream.go
// https://scrapbox.io/data-structures/Realtime_Queue
package main

import "fmt"

func main() {
	s := NewStream()
	s2 := s.Push(1).Push(2).Push(3)
	fmt.Println(s2.Top())
	rev := s2.Reverse()
	fmt.Println(rev.Top())
	fmt.Println(rev)
}

type S int
type Cell struct {
	resolved S
	next     *Stream
}
type Stream struct {
	*Suspension
}

// 惰性求值的流.
func NewStream() *Stream {
	return &Stream{Suspension: NewSuspension()}
}

// 连接两个流。
func Concat(x, y *Stream) *Stream {
	return &Stream{Suspension: NewSuspensionWith(func() interface{} {
		if x.Empty() {
			return y.Resolve()
		}
		return &Cell{resolved: x.Top(), next: Concat(x.Pop(), y)}
	})}
}

func (s *Stream) Empty() bool {
	return s.Resolve() == nil
}

func (s *Stream) Top() S {
	return s.Resolve().(*Cell).resolved
}

func (s *Stream) Pop() *Stream {
	return s.Resolve().(*Cell).next
}

func (s *Stream) Push(x S) *Stream {
	return &Stream{Suspension: NewSuspensionWith(&Cell{resolved: x, next: s})}
}

func (s *Stream) Reverse() *Stream {
	return &Stream{Suspension: NewSuspensionWith(func() interface{} {
		x := s
		res := NewStream()
		for !x.Empty() {
			res = res.Push(x.Top())
			x = x.Pop()
		}
		return res.Resolve()
	})}
}

func (s *Stream) String() string {
	x := s
	res := []S{}
	for !x.Empty() {
		res = append(res, x.Top())
		x = x.Pop()
	}
	for i, j := 0, len(res)-1; i < j; i, j = i+1, j-1 {
		res[i], res[j] = res[j], res[i]
	}
	return fmt.Sprintf("Stream%v", res)
}

// 惰性求值.
type Suspension struct {
	x        interface{}
	resolved interface{}
}

func NewSuspension() *Suspension {
	return &Suspension{}
}

func NewSuspensionWith(x interface{}) *Suspension {
	return &Suspension{x: x}
}

func (s *Suspension) Resolve() interface{} {
	if s.resolved == nil {
		if f, ok := s.x.(func() interface{}); ok {
			s.resolved = f()
		} else {
			s.resolved = s.x
		}
	}
	return s.resolved
}

File: 23_设计类/设计版本控制-可持久化/可持久化数组/PersistentQueue/Stream.ts
/* eslint-disable no-console */
/* eslint-disable @typescript-eslint/no-this-alias */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable eqeqeq */

import { Suspension } from './Suspension'

type Cell<C> = readonly [resolved: C, next: Stream<C>] | undefined | null

/**
 * 惰性求值的流。
 */
class Stream<S> extends Suspension<Cell<S>> {
  /**
   * 连接两个流。
   */
  static concat<T>(x: Stream<T>, y: Stream<T>): Stream<T> {
    return new Stream(() => {
      if (x.empty()) {
        return y.resolve()
      }
      return [x.top()!, Stream.concat(x.pop()!, y)]
    })
  }

  constructor()
  constructor(cell: Cell<S> | (() => Cell<S>))
  constructor(value: S, stream: Stream<S>)
  constructor(arg1?: S | Cell<S> | (() => Cell<S>), arg2?: Stream<S>) {
    const n = arguments.length // 实参个数
    if (n === 2) {
      super([arg1 as S, arg2 as Stream<S>])
    } else if (n === 1) {
      super(arg1 as Cell<S> | (() => Cell<S>))
    } else {
      super()
    }
  }

  empty(): boolean {
    return !this.resolve()
  }

  /**
   * 返回流的第一个元素。
   */
  top(): S | undefined {
    const res = this.resolve()
    return res ? res[0] : undefined
  }

  /**
   * 返回一个流，该流将第一个元素弹出。
   */
  pop(): Stream<S> | undefined {
    const res = this.resolve()
    return res ? res[1] : undefined
  }

  /**
   * 返回一个流，该流将元素`x`推送到前面。
   */
  push(x: S): Stream<S> {
    return new Stream(x, this)
  }

  /**
   * 返回流的反向流。
   */
  reverse(): Stream<S> {
    return new Stream(() => {
      let x: Stream<S> = this
      let res = new Stream<S>()
      while (!x.empty()) {
        res = res.push(x.top()!)
        x = x.pop()!
      }
      return res.resolve()
    })
  }

  override toString(): string {
    let x: Stream<S> = this
    const res: S[] = []
    while (!x.empty()) {
      res.push(x.top()!)
      x = x.pop()!
    }
    res.reverse()
    return `Stream{${res.join(', ')}}`
  }
}

export { Stream }

if (require.main === module) {
  const stream = new Stream<number>()
  const stream2 = stream.push(1).push(2).push(3)
  console.log(stream2.top())
  const reversed = stream2.reverse()
  console.log(reversed.top(), 111)
  console.log(reversed.toString())
}
